### ISSUE TITLE
System.Text.Json - Allow Deserialisation to set properties with private setters without polluting domain objects with deserialisation attributes

### ISSUE THREAD

**Issue by richstokoe** [dotnet/runtime#97206]
### Background and motivation

TL;DR - Could we please have a new option added to JsonSerializerOptions that instructs it to set properties on a model which has non-public setters without requiring the JsonInclude attribute?

Problem:
I welcome corrections, but I understand that with System.Text.Json, it is not possible to tell the deserializer to set properties with private setters. This used to be [possible with Newtonsoft](https://mynkow.github.io/How-to-use-private-setters-with-NewtonSoftJson/) using `BindingFlags.NonPublic` on `DefaultContractResolver.DefaultMemberSearchFlags`  but seems to have become [obseleted](https://www.newtonsoft.com/json/help/html/P_Newtonsoft_Json_Serialization_DefaultContractResolver_DefaultMembersSearchFlags.htm).

This was very useful for designing domain objects that could be serialised to a data store and back, but ensuring that once those objects had been rehydrated, those properties could only be mutated via public methods with validation steps etc.

System.Text.Json supports this with JsonInclude attributes on the model, but I don't want to pollute my domain layer with knowledge of externalities like data stores and serialization. This is poor separation of concerns.

### API Proposal

```csharp 

public class JsonSerializerOptions
{
    // If true, will set properties that have private, protected or internal setters. Defaults to False (for back-compat)
    public bool IncludePropertiesWithNonPublicSetters {get; set;}
}
```

### API Usage

```csharp

// Person type with Name property that is only settable via a public method
// which performs validation
public class Person
{
  public string Name { get; private set; }
  public void SetName(string newName)
  {
    ArgumentException.ThrowIfNullOrEmpty(newName);
    this.Name = newName;
  }
}

// ... 

// Deserializing a serialised Person with a name property will rehydrate this property
var person = System.Text.Json.JsonSerializer.Deserialize<Person>(@"{  "Name": "Bob" }", new JsonSerializerOptions()
{
    IncludePropertiesWithNonPublicSetters = true
);   
```

### Risks

If this property was true by default, it would be a breaking change. It is important that this property is opt-in.

**Comment by eiriktsarpalis** [comment]
It's not something that we would recommend or consider offering as a global setting. Have you considered using a contract customizer to achieve this goal? 

```C#
static void AppendNonPublicSetters(JsonTypeInfo typeInfo)
{
    if (typeInfo.Kind is not JsonTypeInfoKind.Object)
    {
        return;
    }

    foreach (JsonPropertyInfo jsonPropInfo in typeInfo.Properties)
    {
        if (jsonPropInfo.Set is null &&
            jsonPropInfo.AttributeProvider is PropertyInfo { SetMethod: MethodInfo setter })
        {

            jsonPropInfo.Set = (obj, value) => setter.Invoke(obj, [value]);
        }
    }
}
```

It can be used as follows:

```C#
var options = new JsonSerializerOptions
{
    TypeInfoResolver = new DefaultJsonTypeInfoResolver
    {
        Modifiers = { AppendNonPublicSetters }
    }
};

var result = JsonSerializer.Deserialize<MyPoco>("{\"Value\": 42}", options);
Console.WriteLine(result.Value); // 42

public class MyPoco
{
    public int Value { get; private set; }
}
```
